From 6369dd4863619eb00fb26524228903b29b97dd96 Mon Sep 17 00:00:00 2001
From: Leonard Crestez <leonard.crestez@nxp.com>
Date: Mon, 17 Sep 2018 16:42:14 +0300
Subject: [PATCH 5054/5242] MLK-20181-6: drm/mxsfb: Add PM_SLEEP support

commit  5f5c92ee0a6f49b0dddf36addd60cf45cc18964b from
https://source.codeaurora.org/external/imx/linux-imx.git

Since power to the lcdif block can be lost on suspend implement
PM_SLEEP_OPS using drm_mode_config_helper_suspend/resume to save/restore
the current mode.

Signed-off-by: Leonard Crestez <leonard.crestez@nxp.com>
Reviewed-by: Stefan Agner <stefan@agner.ch>
Reviewed-by: Sean Paul <seanpaul@chromium.org>
Signed-off-by: Stefan Agner <stefan@agner.ch>
Link: https://patchwork.freedesktop.org/patch/msgid/cfa1a4083eefd112362e640deeb2e120584ac3f5.1537191359.git.leonard.crestez@nxp.com
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/mxsfb/mxsfb_drv.c |   52 +++----------------------------------
 1 file changed, 4 insertions(+), 48 deletions(-)

diff --git a/drivers/gpu/drm/mxsfb/mxsfb_drv.c b/drivers/gpu/drm/mxsfb/mxsfb_drv.c
index 051a2b5..94bd8c4 100644
--- a/drivers/gpu/drm/mxsfb/mxsfb_drv.c
+++ b/drivers/gpu/drm/mxsfb/mxsfb_drv.c
@@ -542,67 +542,23 @@ static int mxsfb_remove(struct platform_device *pdev)
 	return 0;
 }
 
-#ifdef CONFIG_PM
-static int mxsfb_runtime_suspend(struct device *dev)
-{
-	struct drm_device *drm = dev_get_drvdata(dev);
-	struct mxsfb_drm_private *mxsfb = drm->dev_private;
-
-	if (!drm->registered)
-		return 0;
-
-	if (mxsfb->enabled) {
-		mxsfb_crtc_disable(mxsfb);
-		mxsfb->suspended = true;
-	}
-
-	return 0;
-}
-
-static int mxsfb_runtime_resume(struct device *dev)
-{
-	struct drm_device *drm = dev_get_drvdata(dev);
-	struct mxsfb_drm_private *mxsfb = drm->dev_private;
-
-	if (!drm->registered || !mxsfb->suspended)
-		return 0;
-
-	mxsfb_crtc_enable(mxsfb);
-	mxsfb->suspended = false;
-
-	return 0;
-}
-
+#ifdef CONFIG_PM_SLEEP
 static int mxsfb_suspend(struct device *dev)
 {
 	struct drm_device *drm = dev_get_drvdata(dev);
-	struct mxsfb_drm_private *mxsfb = drm->dev_private;
-
-	if (mxsfb->enabled) {
-		mxsfb_crtc_disable(mxsfb);
-		mxsfb->suspended = true;
-	}
 
-	return 0;
+	return drm_mode_config_helper_suspend(drm);
 }
 
 static int mxsfb_resume(struct device *dev)
 {
 	struct drm_device *drm = dev_get_drvdata(dev);
-	struct mxsfb_drm_private *mxsfb = drm->dev_private;
-
-	if (!mxsfb->suspended)
-		return 0;
 
-	mxsfb_crtc_enable(mxsfb);
-	mxsfb->suspended = false;
-
-	return 0;
+	return drm_mode_config_helper_resume(drm);
 }
 #endif
 
 static const struct dev_pm_ops mxsfb_pm_ops = {
-	SET_RUNTIME_PM_OPS(mxsfb_runtime_suspend, mxsfb_runtime_resume, NULL)
 	SET_SYSTEM_SLEEP_PM_OPS(mxsfb_suspend, mxsfb_resume)
 };
 
@@ -613,7 +569,7 @@ static int mxsfb_resume(struct device *dev)
 	.driver	= {
 		.name		= "mxsfb_drm",
 		.of_match_table	= mxsfb_dt_ids,
-		.pm = &mxsfb_pm_ops,
+		.pm		= &mxsfb_pm_ops,
 	},
 };
 
-- 
1.7.9.5

